src/librbd/io/ImageRequest.cc

```diff
@@ -511,7 +522,7 @@ template <typename I>
 void ImageDiscardRequest<I>::prune_object_extents(ObjectExtents &object_extents) {
   I &image_ctx = this->m_image_ctx;
   CephContext *cct = image_ctx.cct;
-  if (!cct->_conf->rbd_skip_partial_discard) {
+  if (!cct->_conf->rbd_skip_partial_discard || this->m_force_discard) {
```

Comment:
oh, yes, I didn't find the ImageCtx::skip_partial_discard which is missing, will update to add it.